### PR TITLE
[feat] Support KeyValue Schema.

### PR DESCRIPTION
**Pull request by shibd**
### Motivation

C++ client Support KeyValue Schema. For key and value schema, just only support `AVRO` and `JSON` type(consistent with java client).

This PR has been reviewed in pulsar repo: https://github.com/apache/pulsar/pull/17125

### Modifications

- A new constructor is added in `SchemaInfo` to combine key and value schemas.
- Add a new `KeyValue` class, to help users merge and parse key and value data.

### Documentation

The current Schema documentation is relatively simple, I will open a new PR to add more content and sample code
- [x] `doc-required` 
(Your PR needs to update docs and you will update later)

### Matching PR in forked repository

https://github.com/shibd/pulsar-client-cpp/pull/2

### CODE REVIEW

**Comment by BewareMyPower** [comment]
Could you rebase to master?

**Comment by shibd** [comment]
> Could you rebase to master?

Yes, I rebased it. Can you take a look? Thanks.

File: lib/Schema.cc [hunk 1]
```diff
@@ -90,6 +132,45 @@ SchemaInfo::SchemaInfo(SchemaType schemaType, const std::string &name, const std
...
+    std::string valueSchemaStr = valueSchema.getSchema();
+    uint32_t keySize = keySchemaStr.size();
+    uint32_t valueSize = valueSchemaStr.size();
+
+    auto buffSize = sizeof keySize + keySize + sizeof valueSize + valueSize;
+    SharedBuffer buffer = SharedBuffer::allocate(buffSize);
+    buffer.writeUnsignedInt(keySize == 0 ? -1 : static_cast<uint32_t>(keySize));
+    buffer.write(keySchemaStr.c_str(), static_cast<uint32_t>(keySize));
+    buffer.writeUnsignedInt(valueSize == 0 ? -1 : static_cast<uint32_t>(valueSize));
+    buffer.write(valueSchemaStr.c_str(), static_cast<uint32_t>(valueSize));
```
Comment (BewareMyPower) [inline lib/Schema.cc:147]: It's better to replace magic number -1 with `INVALID_SIZE`.

File: lib/KeyValueImpl.cc [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            buffer.consume(keySize);
+        }
+        auto valueSize = buffer.readUnsignedInt();
+        if (valueSize != INVALID_SIZE) {
+            valueBuffer_ = buffer.slice(0, valueSize);
+        }
+    } else {
+        valueBuffer_ = SharedBuffer::wrap(const_cast<char *>(data), length);
+    }
+}
```
Comment (BewareMyPower) [inline lib/KeyValueImpl.cc:45]: There is a problem that I didn't look deeper, must I just wrote a corner case test.

```c++
    auto schema = SchemaInfo(SchemaInfo(SchemaType::STRING, "String", ""),
                             SchemaInfo(SchemaType::STRING, "String", ""));
    Producer producer;
    client.createProducer(topic, ProducerConfiguration().setSchema(schema), producer);
    producer.send(MessageBuilder().setContent(KeyValue("", "value")).build());

    ConsumerConfiguration conf;
    conf.setSubscriptionInitialPosition(InitialPositionEarliest);
    conf.setSchema(schema);
    Consumer consumer;
    client.subscribe(topic, "sub", conf, consumer);

    Message msg;
    result = consumer.receive(msg, 3000);
    if (result == ResultOk) {
        const auto kv = msg.getKeyValueData();
        LOG_INFO("Received KeyValue(key=" << kv.getKey() << ", value=" << kv.getValueAsString() << ") from "
                                          << msg.getMessageId());
     }
```

Segfault happened. I think the key should be allowed be empty.

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

File: lib/KeyValueImpl.cc (hunk 2 above)
Comment (shibd) [reply lib/KeyValueImpl.cc:45]: I can't reproduce it on my host. 

```
2022-10-26 09:23:37.185 INFO  [0x16fdbb000] ConsumerImpl:231 | [persistent://public/default/test-kv, sub, 0] Created consumer on broker [127.0.0.1:60198 -> 127.0.0.1:6650] 
2022-10-26 09:23:37.200 INFO  [0x100228580] SampleConsumer:50 | Received KeyValue(key=, value=value) from (1,0,-1,0)
2022-10-26 09:23:37.200 INFO  [0x100228580] ClientImpl:517 | Closing Pulsar client with 1 producers and 1 consumers
```

Comment (shibd) [reply lib/KeyValueImpl.cc:45]: >  I think the key should be allowed be empty.

Now the implementation is allowed

File: tests/KeyValueSchemaTest.cc [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+   private:
+    Client client{lookupUrl};
+    std::string jsonSchema =
+        R"({"type":"record","name":"cpx","fields":[{"name":"re","type":"double"},{"name":"im","type":"double"}]})";
+};
+
+TEST_P(KeyValueSchemaTest, testKeyValueSchema) {
+    const std::string topicName = "testKeyValueSchema" + std::to_string(time(nullptr));
```
Comment (BewareMyPower) [inline tests/KeyValueSchemaTest.cc:58]: This test is very flaky. You can run the following commands in local env for multiple times.

```bash
./tests/pulsar-tests --gtest_filter='*KeyValueSchemaTest*'
```

```
/home/xyz/pulsar-client-cpp/tests/KeyValueSchemaTest.cc:81: Failure
Expected equality of these values:
  receiveMsg.getPartitionKey()
    Which is: ""
  keyData
    Which is: "{\"re\":2.1,\"im\":1.23}"
```

Similar issues: https://github.com/apache/pulsar-client-cpp/pull/59. It's better to include `GetParam()` in the topic name. Otherwise, different tests might share the same topic.

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

**Comment by BewareMyPower** [comment]
Overall LGTM. Please fix the flaky test.

File: tests/KeyValueSchemaTest.cc (hunk 3 above)
Comment (shibd) [reply tests/KeyValueSchemaTest.cc]: Fixed.

**Review by BewareMyPower — APPROVED** [review]
